core/src/main/java/org/apache/calcite/rel/metadata/RelMdAllPredicates.java

```diff
@@ -113,19 +115,39 @@ public RelOptPredicateList getAllPredicates(TableScan scan, RelMetadataQuery mq)
   }
 
   /**
-   * Extract predicates for a project.
+   * Extracts predicates for a project.
    */
   public RelOptPredicateList getAllPredicates(Project project, RelMetadataQuery mq) {
     return mq.getAllPredicates(project.getInput());
   }
 
   /**
-   * Add the Filter condition to the list obtained from the input.
+   * Extracts predicates for a Filter.
    */
   public RelOptPredicateList getAllPredicates(Filter filter, RelMetadataQuery mq) {
-    final RelNode input = filter.getInput();
-    final RexBuilder rexBuilder = filter.getCluster().getRexBuilder();
-    final RexNode pred = filter.getCondition();
+    return getAllFilterPredicates(filter, mq, filter.getCondition());
+  }
+
+  /**
+   * Extracts predicates for a Calc.
+   */
+  public RelOptPredicateList getAllPredicates(Calc calc, RelMetadataQuery mq) {
+    final RexProgram rexProgram = calc.getProgram();
+    if (rexProgram.getCondition() != null) {
+      final RexNode condition = rexProgram.expandLocalRef(rexProgram.getCondition());
+      return getAllFilterPredicates(calc, mq, condition);
+    } else {
+      return mq.getAllPredicates(calc.getInput());
+    }
+  }
+
+  /**
+   * Add the Filter condition to the list obtained from the input.
+   */
+  private RelOptPredicateList getAllFilterPredicates(RelNode rel,
+      RelMetadataQuery mq, RexNode pred) {
+    final RelNode input = rel.getInput(0);
```

Comment:
How about just use rel, not rel.getInput(0) ?
And add the doc, the pred comes from the parent of rel.

Reply:
`rel` used in `getAllFilterPredicates` rather than only used for `getInput()`.